[WIP] make_message: Add function to generate valid test messages.

LaPulgaaa opened the pull request:
<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

LaPulgaaa commented:
@zulipbot add "buddy review"

LaPulgaaa commented:
@afeefuddin ready for your review.

LaPulgaaa commented:
@afeefuddin ready for your review!!

gnprice commented on web/tests/lib/make_message.js:
```diff
@@ -0,0 +1,142 @@
+    });
+}
+
+exports.make_server_message = (opts = {}) => {
+    const sender_id = opts.sender_id ?? create_sender_id();
+    const msg_id = opts.id ?? create_msg_id();
+    const msg_type = opts.type ?? DEFAULT_MSG_TYPE;
```
Rather than one function that defaults to either stream/channel messages or DMs (or randomly chooses one or the other as default), I recommend having two functions like `make_channel_message` and `make_direct_message`.

A large fraction of tests that need a message are going to care whether it's a DM or a channel message, so they'll need to specify anyway.

And because the types are effectively different for the two kinds of messages, separating the functions will help a lot in being able to type-check the functions and their callers later, when we eventually make TypeScript possible in our tests. Even before we have a type-checker going, writing the code in a way that a type-checker *could* understand is usually a helpful guideline for writing it in a way that human readers can easily understand.

For an example of how to let those two functions share the parts of their code that make sense to share, see zulip-mobile:
https://github.com/zulip/zulip-mobile/blob/2ae8d8e54dcf25ce760206b0f6fc16923b4375b5/src/__tests__/lib/exampleData.js#L402-L528
and in particular here:
```js
  const baseMessage: StreamMessage = {
    ...messagePropertiesBase,
    ...messagePropertiesFromSender(sender),
    ...messagePropertiesFromStream(streamInner),

    content: '<p>This is an example stream message.</p>',
    content_type: 'text/html',
```

LaPulgaaa replied:
Created two separate function to handle both (`"stream"` and `"private"`) kind of messages. 

>>> For an example of how to let those two functions share the parts of their code that make sense to share, see zulip-mobile:

Thanks for the linking that resource. The newer version tries to address the feedback.

gnprice commented on web/tests/lib/make_message.js:
```diff
@@ -0,0 +1,142 @@
+    return `example_${stream_id}-${recipient_id}`;
+}
+
+function create_reactions(sender_id) {
```
I expanded on this principle a bit in chat just now:
https://chat.zulip.org/#narrow/stream/6-frontend/topic/typescript.20node.20tests/near/1896896

LaPulgaaa replied:
Thanks for the detailed explanation!! I have tried to addressed the feedbacks in the latest version.

LaPulgaaa committed:
make_message: Add helper functions to generate better test data.

LaPulgaaa commented:
Updates since the last review:

1. Created two separate functions - `make_stream_server_msg` & `make_private_server_msg` to handle both kind of messages (`"stream"` and `"private"`) thus addressing https://github.com/zulip/zulip/pull/31006#discussion_r1685102371

2. Addressed https://github.com/zulip/zulip/pull/31006#discussion_r1685089037 
>>> So in particular that means a message should by default have no reactions, no topic links, and definitely no submessages.

by removing any interesting fields like `reactions`, `submessage`s, `topic_links` from the default `message` skelaton unless the caller test functions mention those fields.

>>>These helpers generating a reaction, a topic link, or a submessage can potentially still be useful to have here. The way to provide them is for this file to just export them, and then tests can call them and pass the results to functions like make_server_message.

>>>(Probably they'll be most helpful as versions that make just a single reaction etc., 


Also added helper functions (`make_reaction`, `make_submessage` ....) to generate these fields. This would help test functions generate these objects easily and without much code duplication.

LaPulgaaa commented:
@gnprice Thanks for the review. This is ready for your another look.

gnprice reviewed (commented):
Cool, thanks for the revision. Those changes look good.

One other high-level comment below on the strategy. Then for comments on the details of the code, I'll leave that to your buddy, mentor, and so on.

timabbott commented on tools/test-js-with-node:
```diff
@@ -298,6 +298,7 @@ EXEMPT_FILES = make_set(
         "web/src/zform.js",
         "web/src/zulip.js",
         "web/src/zulip_test.ts",
+        "web/tests/lib/make_message.js",
```
What parts of this file don't have coverage? It is probably better to just add a small `test_make_message` test file to exercise it.

timabbott commented on web/tests/lib/make_message.js:
```diff
@@ -0,0 +1,275 @@
+        type: "stream",
+        local_id: opts.local_id,
+    };
+}
```
I think it'd be valuable for this file to be TypeScript, and return a `Message` or `RawMessage` or whatever the appropriate type is; that really helps to validate your work here.

gnprice replied:
That would be great but that's not an option yet, right? That's my understanding of Anders's comments here:
https://chat.zulip.org/#narrow/stream/6-frontend/topic/typescript.20node.20tests/near/1886768

LaPulgaaa replied:
> That would be great but that's not an option yet, right? That's my understanding of Anders's comments here: https://chat.zulip.org/#narrow/stream/6-frontend/topic/typescript.20node.20tests/near/1886768

Yes. That's my understanding of the discussion too.

timabbott replied:
OK, sounds good. Better to do this than not, so no objection here.

timabbott commented on web/tests/lib/make_message.js:
```diff
@@ -0,0 +1,275 @@
+    return {
+        stream_id,
+        subject: subject ?? `topic:${topic_count}`,
+        display_recipient: display_recipient ?? `channel:${stream_id}`,
```
I guess we now have those methods, but I'd still be fine with merging a rebased version of this that doesn't yet use those helpers; it will be a nontrivial design problem to integrate the helpers.
